### PR TITLE
Feat/4685 support tags on contents

### PR DESCRIPTION
**Pull request by raphj**
#4685 

## Checkpoints

**For developers**

- [X] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [X] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [X] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [X] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [X] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [ ] The code is clear enough
- [ ] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [ ] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [ ] Manual, quality tests have been done

### CODE REVIEW

**Review by inkhey — CHANGES_REQUESTED** [review]

File: backend/tracim_backend/lib/core/tag.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        if tag_id:
+            query = query.filter(Tag.tag_id == tag_id)
+        if content_id:
+            query = query.join(TagOnContent, Tag.tag_id == TagOnContent.tag_id).filter(
+                TagOnContent.content_id == content_id
+            )
+        if workspace_id:
+            query = query.filter(Tag.workspace_id == workspace_id)
+        if tag_name:
+            query = query.filter(Tag.tag_name == tag_name)
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/lib/core/tag.py:44]: Yes, we may need to add an autocomplete feature in the future.

(just to be sure, this is not really a suggestion to commit?)

File: backend/tracim_backend/lib/core/tag.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+from tracim_backend.models.data import Workspace
+from tracim_backend.models.tag import Tag
+from tracim_backend.models.tag import TagOnContent
+from tracim_backend.models.tracim_session import TracimSession
+
+
+class TagLib:
+    def __init__(self, session: TracimSession) -> None:
+        self._session = session
+        pass
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/lib/core/tag.py:19]: indeed, this is copy pasted from reaction.py which also has the issue. I'm fixing this there as well.

File: backend/tracim_backend/lib/core/tag.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (inkhey) [reply backend/tracim_backend/lib/core/tag.py:44]: there is not suggestion to commit here.

File: backend/tracim_backend/tests/functional/test_tags.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+# from tracim_backend.views.core_api.schemas import UserDigestSchema
+
+TAG_URLS = (
+    "/api/workspaces/{workspace_id}/contents/{content_id}/tags",
+    "/api/workspaces/{workspace_id}/tags",
+)
+
+SAMPLE_TAG_LIST = (
+    "mytag",
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/tests/functional/test_tags.py:18]: There is one : "a tag"

File: backend/tracim_backend/models/tag.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+from sqlalchemy.orm import relationship
+from sqlalchemy.types import Integer
+
+from tracim_backend.models.data import Content
+from tracim_backend.models.data import Workspace
+from tracim_backend.models.meta import DeclarativeBase
+from tracim_backend.models.mixins import CreationDateMixin
+
+
+class Tag(CreationDateMixin, DeclarativeBase):
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/models/tag.py:15]: I hesitated to do that, but then a tag may be modified in the future by someone else, how should we handle that?

File: backend/tracim_backend/models/tag.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    workspace = relationship("Workspace", remote_side=[Workspace.workspace_id], lazy="joined")
+
+    def __repr__(self):
+        return "<Tag(workspace_id=%s, tag_name=%s)>" % (
+            repr(self.workspace_id),
+            repr(self.tag_name),
+        )
+
+
+class TagOnContent(DeclarativeBase):
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/models/tag.py:36]: Good idea, I will do this.

File: backend/tracim_backend/tests/functional/test_tags.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        # assert last_event.event_type == "tag.created"
+        # author = web_testapp.get("/api/users/{}".format(admin_user.user_id), status=200).json_body
+        # assert last_event.author == UserDigestSchema().dump(author).data
+        # web_testapp.authorization = ("Basic", ("admin@admin.admin", "admin@admin.admin"))
+        # workspace = web_testapp.get(
+        # "/api/workspaces/{}".format(test_workspace.workspace_id), status=200
+        # ).json_body
+        # assert last_event.workspace == {k: v for k, v in workspace.items() if k != "description"}
+        # assert last_event.tag["tag_id"] == tag_id
+        # assert last_event.tag["tag_name"] == tag_name
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/tests/functional/test_tags.py:175]:
```suggestion
```

File: backend/tracim_backend/tests/functional/test_tags.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        # assert last_event.event_type == "tag.deleted"
+        # author = web_testapp.get("/api/users/{}".format(admin_user.user_id), status=200).json_body
+        # assert last_event.author == UserDigestSchema().dump(author).data
+        # web_testapp.authorization = ("Basic", ("admin@admin.admin", "admin@admin.admin"))
+        # workspace = web_testapp.get(
+        # "/api/workspaces/{}".format(test_workspace.workspace_id), status=200
+        # ).json_body
+        # assert last_event.workspace == {k: v for k, v in workspace.items() if k != "description"}
+        # assert last_event.tag["tag_id"] == tag_id
+        # assert last_event.tag["tag_name"] == tag_name
```
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/tests/functional/test_tags.py:279]:
```suggestion
```

File: backend/tracim_backend/lib/core/tag.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (raphj) [reply backend/tracim_backend/lib/core/tag.py:44]: ok, resolving. I'll let charline know your suggestion.

**Review by inkhey — APPROVED** [review]

**Comment by raphj** [comment]
We merged this in the corresponding frontend branch (https://github.com/tracim/tracim/pull/4688), closing

**Pull request closed without merging**